### ISSUE TITLE
"Sync Now" reports success even if the replication fails due to timeout

### ISSUE THREAD

**Issue by kennsippell** [medic/cht-core#8303]
**Describe the bug**
When user clicks "Sync Now," a request is sent to API which times out if it takes >10 minutes. If this request fails due to timeout, no files are moved and the replication is not successful. Despite the timeout, the user sees a success message: "All reports synced".

**To Reproduce**
1. Login as a heavy user on a congested server. Alternatively, prepare a custom build of cht-core which reduces the timeout from 10minutes to something short (eg. <1 second).
2. Fill some forms on the device which should replicate up. Create some forms on the server which should replicate down.
3. Click "sync now".  
4. The _changes request needs to take longer than the timeout value to reproduce this issue. You can test if the timeout happened because the files you created will not have synced.
5. Observe the sync status in the hamburger menu - you will see "All reports synced" in green.

**Expected behavior**
Users should be see a failure message for syncronizations which timeout.
It is good these timeouts are not retrying on failure - we likely shouldn't change that

**Logs**
`/medic/_changes` requests which fail in this manner return a mostly empty result with `200` status code

**Environment**
- Instance: dev
- Browser: Chrome
- App: webapp
- Version: main (4.3)

**Additional context**
Based on recent analysis, ~10% of requests to `/medic/_changes` timeout in production for Muso-Mali over a period of 23 days.

**Comment by dianabarsan** [comment]
Thanks for reporting. This is an interesting issue. 

Can you please detail where the 10 minute timeout comes from? For Muso, that is on 3.x, the ALB timeout should be 1h, and API should be sending [heartbeats](https://github.com/medic/cht-core/blob/master/webapp/src/ts/services/db-sync.service.ts#L97) every 10 seconds to keep the connection alive, so the timeout should _not_ happen, at least in theory. 

In 4.x, this _would_ happen nginx timeout _was_ 10 minutes, but is now increased to 1h in `master`. 

Can you please share some log excerpts of a request that times out after 10 minutes?

**Comment by kennsippell** [comment]
Here are two log lines for one request and response. The response came 10 minutes after the request. The response_time is 10 seconds (first heartbeat). 

```
[2023-04-16 21:26:24] REQ 6c99612c-329d-46e4-9937-cd7fe9e054d4 1.1.1.1 - GET /medic/_changes?timeout=600000&style=all_docs&heartbeat=10000&since=74668764-g1AAAAJ7eJzLYWBg4MhgTmEQTM4vTc5ISXIwNDLXMwBCwxygFFMiQ5L8____szKYkxj6QhRzgWLsxilpJolpZtj04DEpSQFIJtnDDSsXghiWZpBiapBEqmEOIMPi4YbF7YUYlmxobmiUSKphCSDD6uGGWYdBDDMzTDNNMyLRsDwWIMnQAKSA5s2HGGh8DWxgkrFxqoGpAVkGLoAYuB9iYH4wxEBzUyODZGOyDDwAMfA-xMCsK2ADLUwN01LNzMky8AHEQGgYPp8JCUMjg7TkRFNsWrMA9F6qRA&limit=100 HTTP/1.1
[2023-04-16 21:36:24] RES 6c99612c-329d-46e4-9937-cd7fe9e054d4 1.1.1.1 - GET /medic/_changes?timeout=600000&style=all_docs&heartbeat=10000&since=74668764-g1AAAAJ7eJzLYWBg4MhgTmEQTM4vTc5ISXIwNDLXMwBCwxygFFMiQ5L8____szKYkxj6QhRzgWLsxilpJolpZtj04DEpSQFIJtnDDSsXghiWZpBiapBEqmEOIMPi4YbF7YUYlmxobmiUSKphCSDD6uGGWYdBDDMzTDNNMyLRsDwWIMnQAKSA5s2HGGh8DWxgkrFxqoGpAVkGLoAYuB9iYH4wxEBzUyODZGOyDDwAMfA-xMCsK2ADLUwN01LNzMky8AHEQGgYPp8JCUMjg7TkRFNsWrMA9F6qRA&limit=100 HTTP/1.1 200 - 10815.199 ms
```

Here is some aggregate data from a heavily overloaded production instance. This looks only at request matching the regex `/medic/_changes\?.*timeout=600000.*style.*heartbeat.*since.*limit=100`.

If I take a request like `/medic/_changes?timeout=600000&style=all_docs...` which happens in the browser when I click "Sync Now", I then use Chrome DevTools to `copy as cUrl` and replay it via console.  When I adjust the set the query parameter `timeout=1` I see that the _changes feed results in an "emptyish" result with `200` response code. I've assumed that is also what is happening in production though I am unable to reproduce a 10-minute request. I've opened this because the status code matches, and the req/res are very consistently 10min apart. But perhaps I'm missing something (?)

**Comment by dianabarsan** [comment]
Ok, I'll verify this, but my best guess is that the [timeout](https://github.com/medic/cht-core/blob/master/webapp/src/ts/services/db-sync.service.ts#L98) that we set client-side in pouchdb is the one that cancels these requests.

**Comment by dianabarsan** [comment]
> Server-side we should be able to maintain an "hour long request".

This is not accurate, because we impose the `timeout` property that is passed with the request at API level: https://github.com/medic/cht-core/blob/master/api/src/controllers/changes.js#L37

The way this code works is that it sends an empty response at the 10 minute mark. 
```
HTTP/1.1 200 OK
<many \n heartbeats>
{"results":[],"last_seq":"1383521-BLABLABLA"}
```
This code was added when we had users connected to the server at all times via longpoll requests, and the client would restart the request immediately after the previous request had finished. In this context, ending the request after 10 minutes is no issue.

In the context of clients only syncing once every 5 minutes, this becomes problematic specifically because of the successful reported sync status, which in the case of termination by timeout, is untrue. 
The client will attempt syncing after 5 minutes and get another chance at downloading docs from the server, so from a functional perspective this is less of an issue. 

I'm leaving this open for discussion, whether we should address this by changing how we communicate sync status client-side _or_ discuss disabling the timeout _or_ having the timeout send some sort of "invalid" response that will trigger a different response client-side.

**Comment by dianabarsan** [comment]
This should be resolved via https://github.com/medic/cht-core/issues/8296, which has been released in 4.3.0.
